pkg/kubeapiserver/authorizer/config.go

```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
 
 	return union.New(authorizers...), union.NewRuleResolvers(ruleResolvers...), nil
 }
+
+type ruleResolvedAUthorizer interface {
+	Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error)
+	RulesFor(user user.Info, namespace string) ([]authorizer.ResourceRuleInfo, []authorizer.NonResourceRuleInfo, bool, error)
+}
+
+type withNotSyncedReason struct {
+	ruleResolvedAUthorizer
+	informers []cache.SharedInformer
+}
+
+func (az withNotSyncedReason) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	d, reason, err := az.ruleResolvedAUthorizer.Authorize(ctx, a)
+	if err != nil {
+		return d, reason, err
+	}
+
+	if d == authorizer.DecisionNoOpinion && len(reason) == 0 && !az.hasSynced() {
```

Comment:
Should we check `hasSynced` before calling `Authorize`? There might be a race if informers got synched right after we got results from Authorize.

Reply:
done